mp/importer/batcher.py

```diff
@@ -5,32 +5,45 @@
 The user can select the batch size, the maximum number of batches and the starting batch for yielding elements
 Other function will be added to randomly sample elements and to retrieve the last batch  
 """
-def run_in_batches(iterable, end_batch_callback=None, batch_size=2, batch_start=0, max_batches=None):	
+def run_in_batches(iterable, end_batch_callback=None, batch_size=2, batch_start=0, max_batches=None, seed=None, percentage=None):	
     """ Create batches from an iterable
    
     This function is a generator that creates batches of elements from an iterable 
     The callback function is called when a) maximum batches are reached or b) iterable is exhausted 
     All batches (complete and incomplete) are ended by a callback function
+
+    If random options are not None, gets a percentage of the total data
+    Pseudo-samples randomly the elements in the selected percentage (approximate)
+    The main script logs the seed to make it repeatable
     """ 	
+    import random
     c = 0
     ended = False
+    if percentage is not None:									
+        random.seed(seed)									# seed can be passed as kw to be repeatable
+        number = int(((100-percentage)*batch_size)/100)						# e.g. to yield the 30% of batch i mask the 70%
+        mask = set(random.sample(range(batch_size), number))										
+    else:
+        mask = []
     for k in iterable:
         c = c + 1			
         current_batch, position_in_batch = divmod(c, batch_size)
-        if position_in_batch == 0:
-            current_batch = current_batch - 1						# last element of a batch has position=0, value=current_batch+1
-        if current_batch >= batch_start and (max_batches is None or current_batch < batch_start + max_batches):
-            ended = False
-            yield k 
-            if position_in_batch == 0: 							# ends completed batches
-                end_batch_callback()
-                ended = True
-        if max_batches is not None and current_batch == batch_start + max_batches:	# if gets maximum allowed batches, break 
-            break
+        if position_in_batch not in mask:							# yield only values that are not in the mask!
```

Comment:
one of the problems was when there is a 0 in the mask (because element 0 is  to set the current batch), e.g. 
+            if position_in_batch == 0:
+                current_batch = current_batch - 1	

so I replaced 0 with a randomly sampled value:

        if 0 in mask:
            mask.discard(0)
            mask.add(int(random.random()*10))

however, the test for py3 still fails in giving incomplete batches